R/enumerate_all_paths.R

```diff
@@ -0,0 +1,160 @@
+
+  graph_df_ud <-
+    all_fks %>%
+    rename(
+      child_table = parent_table,
+      child_cols = parent_cols,
+      parent_table = child_table,
+      parent_cols = child_cols,
+    ) %>%
+    bind_rows(all_fks, .)
+
+  helper_env <- new_environment()
+  assign("tbl_node", list(), envir = helper_env)
+  assign("all_paths", tibble(
+    child_table = character(),
+    child_cols = new_keys(),
+    parent_table = character(),
+    parent_cols = new_keys(),
+    new_child_table = character(),
+    new_parent_table = character()
+  ), envir = helper_env)
+  enumerate_all_paths_impl(start, graph_df_ud = graph_df_ud, helper_env = helper_env)
+  all_paths <- get("all_paths", envir = helper_env)
+  # need to take into account FKs from unconnected components (graph representation)
+  fks_from_unconnected <- anti_join(
+    all_fks,
+    all_paths,
+    by = c("child_table", "parent_table")
+  ) %>%
+    mutate(new_child_table = child_table, new_parent_table = parent_table)
+  all_paths %>%
+    rename_unique() %>%
+    bind_rows(fks_from_unconnected) %>%
+    split_to_list()
+}
+
+enumerate_all_paths_impl <- function(node,
+                                     node_key_cols = character(),
+                                     former_node = character(),
+                                     former_key_cols = character(),
+                                     path = character(),
+                                     graph_df_ud,
+                                     helper_env) {
+  if (length(path) > 0) {
+    # increase tbl_node[[node]] by 1, return this index in a suffix
+    usage_idx <- inc_tbl_node(node, helper_env)
+    add_path_to_all_paths(
+      graph_df_ud,
+      node,
+      node_key_cols,
+      former_node,
+      former_key_cols,
+      new_former_node = names(path)[[length(path)]],
+      usage_idx,
+      helper_env
+    )
+  } else {
+    usage_idx <- ""
+  }
+
+  path <- c(path, set_names(node, paste0(node, usage_idx)))
+  out <-
+    graph_df_ud %>%
+    filter(child_table == !!node) %>%
+    filter(!(parent_table %in% path)) %>%
+    rename(node = parent_table, node_key_cols = parent_cols, former_node = child_table, former_key_cols = child_cols) %>%
+    select(-on_delete)
+
+  if (nrow(out) == 0) {
+    return()
+  }
+
+  pwalk(out, enumerate_all_paths_impl, path, graph_df_ud, helper_env)
+}
+
+rename_unique <- function(all_paths) {
+  all_names <- bind_rows(
+    select(all_paths, table = child_table, new_table = new_child_table),
+    select(all_paths, table = parent_table, new_table = new_parent_table)
+  ) %>%
+    distinct() %>%
+    arrange(table, new_table) %>%
+    group_by(table) %>%
+    count()
+  left_join(all_paths, rename(all_names, n_c = n), by = c("child_table" = "table")) %>%
+    left_join(rename(all_names, n_p = n), by = c("parent_table" = "table")) %>%
+    mutate(
+      new_child_table = if_else(n_c == 1, child_table, new_child_table),
```

Comment:
A simple lookup in a named vector (which needs to be prepared in a suitable way) might be easier to digest:

```suggestion
      new_child_table = (!!tables)[new_child_table],
```

Reply:
It seems to me you have a different approach in mind than how I implemented it.
In this function (`rename_unique(all_paths)`) I am just doing a little cleanup of the result table, such that tables that were visited only once retain their original name.
At this stage the result table itself already contains all old names, new names (a few of which might be reversed in this function), column names, on_delete.

I feel it might be best if you really work with the code and choose what you deem the best solutions. It's completely fine with me if you choose a very different implementation eventually. I assume you will at least be able to make use of the existing tests for both `enumerate_all_paths()` and `dm_disentangle()`.

Reply:
I think I understand the intent of the code. I'll try to simplify (and `(!!tables)[...]` isn't exactly simple or intuitive...).